server/src/instant/db/cel.clj

```diff
@@ -265,8 +337,621 @@
+                            y))))
+
+(defmethod or-overload-fn [:whereclause :bool]
+  ([_]
+   (fn [[^WhereClause x ^Boolean y]]
+     (or y x))))
+
+(defmethod or-overload-fn [:datakey :bool]
+  ([_]
+   (fn [[^DataKey x ^Boolean y]]
+     (or y x))))
+
+(defmethod or-overload-fn [:bool :bool]
+  ([_]
+   (fn [[^Boolean x ^Boolean y]]
+     (or x y))))
+
+(def or-overloads
+  (global-overload
+   (.getFunction Operator/LOGICAL_OR)
+   (for [arg-1 [:datakey :whereclause :bool]
+         arg-2 [:datakey :whereclause :bool]
+         :let [args [arg-1 arg-2]]]
+     {:overload-id (str "_or_" (clojure-string/join "_"
+                                                    (map type->name args)))
+      :cel-args (map type->cel args)
+      :cel-return-type (case args
+                         ([:whereclause :whereclause]
+                          [:datakey :datakey]
+                          [:datakey :whereclause]
+                          [:whereclause :datakey]) whereclause-cel-type
+                         ([:whereclause :bool]
+                          [:bool :whereclause]
+                          [:datakey :bool]
+                          [:bool :datakey]) SimpleType/DYN
+                         [:bool :bool] SimpleType/BOOL
+                         SimpleType/DYN)
+      :java-args (map type->java args)
+      :impl (get-or-overload-fn args)})))
+
+;; Overloads for `AND`
+;; We overload the existing AND function to handle our custom types
+;; cel won't let us replace the `AND` function unless we rewrite te ast
+
+(defn and-overload-dispatch [args]
+  args)
+
+(defmulti and-overload-fn #'and-overload-dispatch)
+
+(defn get-and-overload-fn
+  "Allows us to define the function in one direction only.
+  [:datakey :bool] executes [:bool :datakey] with the args reversed."
+  [arg-typs]
+  (if-let [forward-method (get-method and-overload-fn arg-typs)]
+    (forward-method arg-typs)
+    (when-let [reverse-method (get-method and-overload-fn (reverse arg-typs))]
+      (let [f (reverse-method (reverse arg-typs))]
+        (fn [args]
+          (f (reverse args)))))))
+
+(defmethod and-overload-fn [:datakey :datakey]
+  ([_]
+   (fn [[^DataKey x ^DataKey y]]
+     (combine-where-clauses :and
+                            (WhereClause. {(.data_key x) true})
+                            (WhereClause. {(.data_key y) true})))))
+
+(defmethod and-overload-fn [:whereclause :whereclause]
+  ([_]
+   (fn [[^WhereClause x ^WhereClause y]]
+     (combine-where-clauses :and x y))))
+
+(defmethod and-overload-fn [:datakey :whereclause]
+  ([_]
+   (fn [[^DataKey x ^WhereClause y]]
+     (combine-where-clauses :and
+                            (WhereClause. {(.data_key x) true})
+                            y))))
+
+(defmethod and-overload-fn [:whereclause :bool]
+  ([_]
+   (fn [[^WhereClause x ^Boolean y]]
+     (and y x))))
+
+(defmethod and-overload-fn [:datakey :bool]
+  ([_]
+   (fn [[^DataKey x ^Boolean y]]
+     (and y x))))
```

Comment:
Neat! So this effectively "passes" through the DataKey, to the next thing, which may then convert it into a where clause. Mind is being blown 🤯